### ISSUE TITLE
Cannot reproduce results 

### ISSUE THREAD

**Issue by guillembraso** [muyangyi/SimSeg#10]
Hi,

I am trying to reproduce your results. I am using all the hyperparameters reported in your paper's supplementary material, and I am tuning other hyperparameters that you did not specify (e.g. gradient clipping, stochastic depth probability, warmup lr ratio). However, my validation mIoU ends up very off, at around 48.5 on PASCAL VOC for my best runs, with ViT-S, instead of the reported 56.6. 

As I mentioned in my other issue, the training hyperparameters you specify in your config file do not match the ones provided in the paper. So could you please fix that, or provide a list of hyperparameters and relevant details on how to train your model?

Thanks,

Guillem

**Comment by muyangyi** [comment]
This issue is likely related to the pooling strategy, not the hyperparameters.  
Please use `VanillaTopKPooling` when training, and use `TopKPooling` at evaluation. (around line 35 of `simseg/models/pipelines/clip.py`)
`VanillaTopKPooling` and `TopKPooling` can be found in: `simseg/models/components/pooling.py`
Here's the list of hyperparameters when launch training, FYI.
```
data.cuda_eval=False 
data.batch_size=4096 
data.batch_size_train=1024
data.batch_size_val=1024
data.enable_valid=True 
epoch=20
optim.param.weight_decay=0.2
data.train_name=[cc3m,cc12m]
data.train_frac=[1,1]
model.max_length=25
transforms.train_transforms=[random_resize_crop,autoaug]
loss.temperature.name=parameter
data.train_type=batch_shuffle
runner.val_interval_steps=100
loss.nce_loss.gather_backward=True
optim.lr.init=3e-4
ckpt.step_interval=100
optim.lr.name=cosine_schedule_with_warmup_min_lr_scale
optim.lr.param.min_lr_scale=0.1
runner.name=clip_bsgs
runner.stable_random=step
data.train_executor_num=1
data.num_workers=2
loss.name=NCE
model.image_encoder.pretrained=False
model.text_encoder.pretrained=False
loss.temperature.value=0.07
model.image_encoder.name=vit_modelzoo
model.image_encoder.tag=vit_small_patch16_224_in21k
model.image_encoder.embedding_dim=384
model.image_encoder.vit.only_cls_token=False
model.text_encoder.only_cls_token=False
model.pool.name=topk_old
model.pool.topkpooling1d.image_k=5
model.pool.topkpooling1d.text_k=5
dist.find_unused_parameters=False
```

**Comment by guillembraso** [comment]
Hi!

Thanks so much for your reply and sharing additional hyperparameters, it's highly appreciated!

I implemented the change you suggested, and also set `model.pool.topkpooling1d.text_k=5` during training, while keeping it `model.pool.topkpooling1d.text_k=5` during evaluation. This yielded a large improvement from `48.5` on Pascal to `53.2`. It is though still a bit far from the `56+` yielded by the checkpoints you provide, so I believe something is still missing.

Also, `VanillaTopKPooling` only seems to differ from `TopKPooling` in that you use attention masks for text tokens during training. Is this expected to make a big difference? 

Could you please share the full configuration you use during training? It's ok to include irrelevant entries, I'll just ignore them. That'd help me a lot

Some additional things I am not sure about:
- Do you use stochastic depth in the ViT model (i.e. `drop_path_rate>0`
- Do you apply any gradient clipping?
- Do you modify `eps` in the `AdamW` optimizer? Seems to be set to `1e-6`
- Do you set any special param groups in your optimizer (e.g. no weight decay for norm layers, etc.?)

Do you have any other insights on things that could be affecting results?

Thanks,

Guillem

**Comment by guillembraso** [comment]
Also, do you change any hyperparameter settings when switching from `ViT-S` to `ViT-B` ?

**Comment by muyangyi** [comment]
1. Also, `VanillaTopKPooling` only seems to differ from `TopKPooling` in that you use attention masks for text tokens during training. Is this expected to make a big difference?
Yes, by employing `VanillaTopKPooling`, we prevent the text encoder from being "lazy" by compelling it to focus its attention on key entity words. This results in more accurate feature representations by the text encoder during evaluation.

2. The complete configuration is structured in three levels: `simseg/tasks/clip/config.py` < `configs/clip/simseg.vit-s.yaml` < the parameters list above. The latter ones overwrite the former ones. 
You already have the full configuration we used.

3. The rest improvement probably comes from the evaluation parameters, they are quite fragile. We have other checkpoints that only achieve `53+ mIoU` with completely same training and evaluation. Please try the following params in `tools/seg_evaluation.py`: 
   - dense crf params (line 48 - 51)
   - dilate and erode (line 157 - 159)
   - threshold (line 124)
   - top_cls_num (line 122)

5. For the additional things:
   - no stochastic depth nor change `drop_path_rate`
   - no gradient clipping
   - `eps` is `1e-6`
   - not setting any special param groups
   - `ViT-S` and `ViT-B` use same training params

**Comment by muyangyi** [comment]
> while keeping it `model.pool.topkpooling1d.text_k=5` during evaluation.

Not sure if this is typo, but during evaluation `text_k` should be 1.

**Comment by guillembraso** [comment]
Thanks a lot for your answers. I am very puzzled by them though, if I use the settings you described (specifically `drop_path_rate=0` and no special parameter groups in my optimizer, my evaluation mIoU drops dramatically, to around `38`, so I believe I might be missing something. Will dig into it and get back to you if I find out anything. Will close this issue for now.

**Comment by ByeongJuWoo** [comment]
@guillembraso Hi, I'm also interested in reproducing this work. Could you give me an advice for the hyperparameter setting you found? I want to reproduce this work on openclip framework. Thanks a lot!!
